src/packet/public_key_encrypted_session_key.rs

```diff
@@ -47,61 +61,181 @@ impl PublicKeyEncryptedSessionKey {
     ) -> Result<Self> {
         // the session key is prefixed with symmetric key algorithm
         let len = session_key.len();
-        let mut data = vec![0u8; len + 3];
+        let mut data = vec![0u8; len + 1];
         data[0] = u8::from(alg);
         data[1..=len].copy_from_slice(session_key);
 
-        // and appended a checksum
-        BigEndian::write_u16(
-            &mut data[len + 1..],
-            checksum::calculate_simple(session_key),
-        );
+        // Appended a checksum, except for x25519/x448
+        // FIXME: factor this difference out and up?
+        match pkey.public_params() {
+            PublicParams::X25519 { .. } => {}
+            _ => {
+                // and appended a checksum
+                data.extend_from_slice(&checksum::calculate_simple(session_key).to_be_bytes())
+            }
+        }
 
-        let mpis = pkey.encrypt(rng, &data)?;
+        let values = pkey.encrypt(rng, &data, false)?;
 
-        Ok(PublicKeyEncryptedSessionKey {
+        Ok(PublicKeyEncryptedSessionKey::V3 {
             packet_version: Default::default(),
-            version: 3,
             id: pkey.key_id(),
-            algorithm: pkey.algorithm(),
-            mpis,
+            pk_algo: pkey.algorithm(),
+            values,
         })
     }
 
+    /// Encrypts the given session key to the passed in public key, as a v6 pkesk.
+    /// FIXME: cleanup/DRY with from_session_key
+    pub fn from_session_key6<R: CryptoRng + Rng>(
+        rng: &mut R,
+        session_key: &[u8],
+        pkey: &impl PublicKeyTrait,
+    ) -> Result<Self> {
+        // the session key
+        let mut data = session_key.to_vec();
+
+        // Appended a checksum, except for x25519/x448
+        // FIXME: factor this difference out and up?
+        match pkey.public_params() {
+            PublicParams::X25519 { .. } => {}
+            _ => data.extend_from_slice(&checksum::calculate_simple(session_key).to_be_bytes()),
+        }
+
+        let values = pkey.encrypt(rng, &data, true)?;
+
+        Ok(PublicKeyEncryptedSessionKey::V6 {
+            packet_version: Default::default(),
+            key_version: Some(pkey.version()),
+            fingerprint: Some(pkey.fingerprint()),
+            pk_algo: pkey.algorithm(),
+            values,
+        })
+    }
+
+    // FIXME: helper fn to check if a Key matches with this PKESK's target (can we remove id/fp?)
+    // v3: keyid or wildcard match
+    // v6: fp or implicit wildcard match
+
+    pub fn match_identity(&self, key_id: &KeyId, fp: &Fingerprint) -> bool {
+        match self {
+            Self::V3 { id, .. } => {
+                let wildcard_id =
+                    KeyId::from_slice(&[0, 0, 0, 0, 0, 0, 0, 0]).expect("wildcard_id");
```

Comment:
should be a function on `KeyId`, `fn is_wildcard(&self) -> bool`

Reply:
good point (ae28f112b8ea74d6ca7070bd330e9d08506a5e2c)